app/models/spree/payment_processing_decorator.rb

```diff
@@ -55,6 +55,29 @@ def set_proper_state(current_state, response, action)
         current_state
       end
     end
+
+    def handle_payment_preconditions
+      unless block_given?
+        raise ArgumentError, 'handle_payment_preconditions must be called with a block'
+      end
+
+      if payment_method&.source_required?
+        if source
+          unless processing?
+            if payment_method.supports?(source) || token_based?
+              yield
+            else
+              invalidate!
+              raise Core::GatewayError, Spree.t(:payment_method_not_supported)
+            end
+          end
+        else
+          raise Core::GatewayError, Spree.t(:payment_processing_failed)
+        end
+      else
+        yield
```

Comment:
This change is going to affect every payment. Do we really need it here and is it safe?

Reply:
That was one of the main problems why the braintree payments were not processed - because at that point they don't have the source so the given block was never actually processed (that means not running the `process_authorization` and `process_purchase` methods - the order was stuck in the payment step).
I've put the yield block more as an idea to discuss possible fixes because, as you said it will affect every processed payment and it's quite important part of the spree/core.
For now i added guard clause to return if there's no payment method but i'm thinking that instead of else, maybe i should add `elsif` and check if the Payment Method is either DropIn, HostedFields or PaypalExpress - for sure it would be safer but i'm not sure what exactly would be best in this case.

What do you think about the `elsif` or maybe adding some other conditionals to this part?

Reply:
I think it should be called only in specific conditions you are talking about.
Also, there is no need to duplicate the whole method. Maybe something like:
```
def handle_payment_preconditions
  unless block_given?
    raise ArgumentError, 'handle_payment_preconditions must be called with a block'
  end

  CONDITION ? yield : super
end
```
could work?

Reply:
Added condition where i check if the `provider` is equal to `Braintree` - that should be enough to make sure it will be fired only for Braintree payment methods :) Thanks for the suggeestion!
I've tested on few basic checkout scenarios and it looks like there's no issues with the functionality, everything works ok